packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts

```diff
@@ -0,0 +1,77 @@
+import rule from '../../src/rules/no-unused-expressions';
+import { RuleTester } from '../RuleTester';
+
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 6,
+    sourceType: 'module',
+    ecmaFeatures: {},
+  },
+  parser: '@typescript-eslint/parser',
+});
+
+// the base rule doesn't have messageIds
+function error(
+  messages: { line: number; column: number }[],
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): any[] {
+  return messages.map(message => ({
+    ...message,
+    message:
+      'Expected an assignment or function call and instead saw an expression.',
+  }));
+}
+
+ruleTester.run('no-unused-expressions', rule, {
+  valid: [
+    `
+      test.age?.toLocaleString();
+    `,
+  ],
+  invalid: [
+    {
+      code: `
+if(0) 0
+        `,
+      errors: error([
+        {
+          line: 2,
+          column: 7,
+        },
+      ]),
+    },
+    {
+      code: `
+f(0), {}
+        `,
+      errors: error([
+        {
+          line: 2,
+          column: 1,
+        },
+      ]),
+    },
+    {
+      code: `
+a, b()
+        `,
+      errors: error([
+        {
+          line: 2,
+          column: 1,
+        },
+      ]),
+    },
+    {
+      code: `
+a() && function namedFunctionInExpressionContext () {f();}
+        `,
+      errors: error([
+        {
+          line: 2,
+          column: 1,
+        },
+      ]),
+    },
+  ],
```

Comment:
You'll need to add some invalid test cases for optional chaining as well.


In particular you need to test cases like:
```ts
a?.b
(a?.b).c

a?.['b']
(a?.['b']).c

a?.b()?.c
(a?.b()).c
```

Have a look at some of the examples in the parser fixtures to get a sense for all of the cases that exist in optional chaining:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-call-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-call.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-element-access-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-element-access.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain-with-parens.src.ts
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/shared-fixtures/fixtures/typescript/basics/optional-chain.src.ts

Reply:
Done